File: BST/ValidateBST/valid.py
# =========================
# Solution 1 : O(n²)
# =========================

class Solution:
    def isMaxTree(self, root, value):
        if not root:
            return True

        if root.data >= value:
            return False

        return (self.isMaxTree(root.left, value) and
                self.isMaxTree(root.right, value))

    def isMinTree(self, root, value):
        if not root:
            return True

        if root.data <= value:
            return False

        return (self.isMinTree(root.left, value) and
                self.isMinTree(root.right, value))

    def isBST(self, root):
        if not root:
            return True

        if (not self.isMaxTree(root.left, root.data) or
            not self.isMinTree(root.right, root.data)):
            return False

        return self.isBST(root.left) and self.isBST(root.right)


# =========================
# Solution 2 : O(n)
# =========================

class Solution:
    def helper(self, root, low, high):
        if not root:
            return True

        if root.data <= low or root.data >= high:
            return False

        return (self.helper(root.left, low, root.data) and
                self.helper(root.right, root.data, high))

    def isBST(self, root):
        return self.helper(root, float('-inf'), float('inf'))

File: BST/ValidateBST/valid.cpp
#include <climits>

class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = right = nullptr;
    }
};

// =========================
// Solution 1 : O(n²)
// =========================

class Solution {
public:
    bool isMaxTree(Node* root, int value) {
        if (!root)
            return true;

        if (root->data >= value)
            return false;

        return isMaxTree(root->left, value) &&
               isMaxTree(root->right, value);
    }

    bool isMinTree(Node* root, int value) {
        if (!root)
            return true;

        if (root->data <= value)
            return false;

        return isMinTree(root->left, value) &&
               isMinTree(root->right, value);
    }

    bool isBST(Node* root) {
        if (!root)
            return true;

        if (!isMaxTree(root->left, root->data) ||
            !isMinTree(root->right, root->data))
            return false;

        return isBST(root->left) &&
               isBST(root->right);
    }
};


// =========================
// Solution 2 : O(n)
// =========================

class Solution {
public:
    bool helper(Node* root, long long low, long long high) {
        if (!root)
            return true;

        if (root->data <= low || root->data >= high)
            return false;

        return helper(root->left, low, root->data) &&
               helper(root->right, root->data, high);
    }

    bool isBST(Node* root) {
        return helper(root, LLONG_MIN, LLONG_MAX);
    }
};